From 888769dbaa5b2fb195e38ff295bb38c9c1f5977c Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Tue, 12 Jul 2022 10:52:09 +0200
Subject: [PATCH 0588/1695] drivers: regulator: update vrefbuf driver for
 STM32MP25

Update vrefbuf driver for STM32MP25.
The only difference is PWR in STM32MP25 support only 2 voltages
but in ascending values in VRS bit-field.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git d11ff39cc0b526825cd088308733a2715b252510 drivers: regulator: update vrefbuf driver for STM32MP25]
Change-Id: I909e640b8ae9608f5ec32fb31ac4b3a086088cd7
Signed-off-by: Pascal Paillet <p.paillet@st.com>
Signed-off-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/341943
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/359154
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 core/drivers/regulator/stm32_vrefbuf.c | 29 ++++++++++++++++++++++----
 1 file changed, 25 insertions(+), 4 deletions(-)

diff --git a/core/drivers/regulator/stm32_vrefbuf.c b/core/drivers/regulator/stm32_vrefbuf.c
index 18ccb099b..cbecd443c 100644
--- a/core/drivers/regulator/stm32_vrefbuf.c
+++ b/core/drivers/regulator/stm32_vrefbuf.c
@@ -23,7 +23,6 @@
 /* VRS bit 3 is unused because the voltage is not specified */
 #define VREFBUF_CSR_VRS			GENMASK_32(5, 4)
 #define VREFBUF_CSR_VRS_SHIFT		U(4)
-#define INV_VRS(x)			((~(x)) & VREFBUF_CSR_VRS)
 
 #define VREFBUF_CSR_VRR			BIT(3)
 #define VREFBUF_CSR_HIZ			BIT(1)
@@ -40,6 +39,7 @@
  */
 struct vrefbuf_compat {
 	int voltages[VREFBUF_LEVELS_COUNT];
+	bool invert_voltages;
 };
 
 /*
@@ -67,6 +67,7 @@ static const struct vrefbuf_compat stm32mp15_vrefbuf_compat = {
 		/* Matches resp. VRS = 011b, 010b, 001b, 000b */
 		1500000, 1800000, 2048000, 2500000,
 	},
+	.invert_voltages = true,
 };
 
 static const struct vrefbuf_compat stm32mp13_vrefbuf_compat = {
@@ -74,6 +75,15 @@ static const struct vrefbuf_compat stm32mp13_vrefbuf_compat = {
 		/* Matches resp. VRS = 011b, 010b, 001b, 000b */
 		1650000, 1800000, 2048000, 2500000,
 	},
+	.invert_voltages = true,
+};
+
+static const struct vrefbuf_compat stm32mp25_vrefbuf_compat = {
+	/* Matches resp. VRS = 000b, 001b */
+	.voltages = {
+		1210000, 1500000,
+	},
+	.invert_voltages = false,
 };
 
 /* Expect at most 1 instance */
@@ -169,8 +179,12 @@ static TEE_Result vrefbuf_get_voltage(struct regulator *regulator,
 	if (res)
 		return res;
 
-	index = io_read32(vr->base + VREFBUF_CSR) & VREFBUF_CSR_VRS;
-	index = INV_VRS(index) >> VREFBUF_CSR_VRS_SHIFT;
+	if (vr->compat->invert_voltages)
+		index = ((~io_read32(vr->base)) & VREFBUF_CSR_VRS) >>
+			VREFBUF_CSR_VRS_SHIFT;
+	else
+		index = (io_read32(vr->base) & VREFBUF_CSR_VRS) >>
+			VREFBUF_CSR_VRS_SHIFT;
 
 	clk_disable(vr->clock);
 
@@ -187,7 +201,10 @@ static TEE_Result vrefbuf_set_voltage(struct regulator *regulator, int level_uv)
 
 	for (i = 0 ; i < ARRAY_SIZE(vr->compat->voltages) ; i++) {
 		if (vr->compat->voltages[i] == level_uv) {
-			uint32_t val = INV_VRS(i << VREFBUF_CSR_VRS_SHIFT);
+			uint32_t val = i << VREFBUF_CSR_VRS_SHIFT;
+
+			if (vr->compat->invert_voltages)
+				val = (~val) & VREFBUF_CSR_VRS;
 
 			res = clk_enable(vr->clock);
 			if (res)
@@ -380,6 +397,10 @@ static const struct dt_device_match stm32_vrefbuf_match_table[] = {
 		.compatible = "st,stm32mp13-vrefbuf",
 		.compat_data = &stm32mp13_vrefbuf_compat
 	},
+	{
+		.compatible = "st,stm32mp25-vrefbuf",
+		.compat_data = &stm32mp25_vrefbuf_compat,
+	},
 	{ }
 };
 
-- 
2.47.2

